packages/astro/src/assets/fonts/utils.ts

```diff
@@ -111,3 +112,76 @@ export function proxyURL({ value, hashString, collect }: ProxyURLOptions): strin
 	// Now that we collected the original url, we return our proxy so the consumer can override it
 	return url;
 }
+
+export function isGenericFontFamily(str: string): str is keyof typeof DEFAULT_FALLBACKS {
+	return Object.keys(DEFAULT_FALLBACKS).includes(str);
+}
+
+type FontFaceMetrics = Parameters<typeof fontaine.generateFontFace>[0];
+
+/**
+ * Generates CSS for a given family fallbacks if possible.
+ *
+ * It works by trying to get metrics (using fontaine) of the provided font family.
+ * If some can be computed, they will be applied to the eligible fallbacks to match
+ * the original font shape as close as possible.
+ */
+export async function generateFallbacksCSS({
+	family,
+	fallbacks: _fallbacks,
+	fontURL,
+	getMetricsForFamily,
+	// eslint-disable-next-line @typescript-eslint/no-shadow
+	generateFontFace,
+}: {
+	/** The family name */
+	family: string;
+	/** The family fallbacks */
+	fallbacks: Array<string>;
+	/** A remote url or local filepath to a font file. Used if metrics can't be resolved purely from the family name */
+	fontURL: string | null;
+	getMetricsForFamily: (family: string, fontURL: string | null) => Promise<null | FontFaceMetrics>;
+	generateFontFace: typeof fontaine.generateFontFace;
+}): Promise<null | { css: string; fallbacks: Array<string> }> {
+	// We avoid mutating the original array
+	let fallbacks = [..._fallbacks];
+	if (fallbacks.length === 0) {
+		return null;
+	}
+
+	let css = '';
+	const metrics = await getMetricsForFamily(family, fontURL);
+	// If there are no metrics, we can't generate useful fallbacks
+	if (!metrics) {
+		return { css, fallbacks };
+	}
+
+	// TODO: to be documented
+	// The last element of the fallbacks is usually a generic family name (eg. serif)
+	const lastFallback = fallbacks[fallbacks.length - 1];
+	// If it's not a generic family name, we can't infer local fonts to be used as fallbacks
+	if (!isGenericFontFamily(lastFallback)) {
+		return { css, fallbacks };
+	}
```

Comment:
Shouldn't this be an error or a warning? I suppose `fallbacks` is passed by the user, so if the last font isn't a generic one, maybe they should be notified

Reply:
I updated the RFC with more info about this but TLDR is it's fine not to specify a generic font family so I don't think we should log anything